### DISCUSSION TITLE
updateTrigger does not work with react state

### DISCUSSION THREAD

**Discussion by hoang-ta** [visgl/deck.gl#5989] [Q&A] [answered]
Hello everyone, 

I am using `updateTrigger` to **change colors of layers randomly every 1 second**  in a React app. 
- My `layer` declaration:
```
  const layer = new GeoJsonLayer({
    ...
    getLineColor: (feature: any) => [0 * trigger, 0 * trigger, 0 * trigger],
    updateTrigger: {
      getLineColor: trigger
    }
  });
```
- Randomly change color based on `trigger` value every 1 second:
```
 const [trigger, setTrigger] = React.useState(0.5);
  React.useEffect(() => {
    setTimeout(() => {
      let newTrigger = Math.random();
      while (newTrigger === trigger) {
        newTrigger = Math.random();
      }
      setTrigger(newTrigger);
    }, 1000);
  }, [trigger]);

```
- The above works fine **but** when I use layer as state, the `updateTrigger` does not work anymore
```
  const [layerState, setLayerState] = React.useState(layer);
```
- [Sandbox url](https://codesandbox.io/s/jovial-pare-dlmqo?file=/src/App.tsx)
- Demo clip:
![update-trigger](https://user-images.githubusercontent.com/61722520/126136726-95437c9f-710f-459f-b2fc-1448688acd91.GIF)
- Here, as you can see, the `updateTrigger` only works with `layer` (**normal variable**) but does not work with `layerState` ( **React state**)

As my source code is using all layers as states and now i start using `updateTrigger` for my layer so i need to find a way for to solve this without removing codes involving my states.

Appreciate your help :smile:

**Comment by Pessimistress** [comment, marked as answer]
Not sure what you are trying to accomplish by using state here.  But when you call `useState(layer)`, `layer` is the initial state, i.e. the value is only read the first time it's called. After that you need to call `setLayerState` to change `layerState`.

https://reactjs.org/docs/hooks-state.html#declaring-a-state-variable

**Reply by hoang-ta** [reply]
So then i don't need to use the `updateTrigger`, do i ?. Because as the `trigger` state changes, it will re-render the app. Then i have a question about which  case to use `updateTrigger` to take advantage of it ?

- At the moment, i myself understand that `updateTrigger` will **_some how  re-render the deck.gl_** with a new value for the accessor when the value of variable (`trigger`) provided for the accessor (`getLineColor`) change.
- To clarify my confuse, i have updated the code in the [sandbox](https://codesandbox.io/s/jovial-pare-dlmqo?file=/src/App.tsx). 
- Now `trigger` is not a state anymore, i change its value through clicking a button at the bottom of the page

**Reply by Pessimistress** [reply]
You do need to use updateTriggers. Your question was "The above works fine but when I use layer as state, the updateTrigger does not work anymore" and I'm commenting on your `useState(layer)`.

You should read the React documentation I linked. This is not related to deck.gl.

**Reply by hoang-ta** [reply]
Yes i understand now that i need to update `layerState` by using React `setLayerState` hook function in order to change the color without a need to use `updateTriggers`. 

But can you help correct my understand about `updateTriggers`, when and how to use it properly? Here is the following sample in DeckGl document: 

```
  const layer = new ScatterplotLayer({
    ... // Other props
    getFillColor: d => d.male ? maleColor : femaleColor,
    updateTriggers: {
        getFillColor: [maleColor, femaleColor]
    }
  });
```
> updateTriggers expect an object whose keys are names of accessor props of this layer, and values are one or more variables that affect the output of the accessors.
> For example, updateTriggers.getFillColor is a list of variables that affect the output of getFillColor. If either value in the array changes, all attributes that depend on getFillColor will be updated. The variables may be numbers, strings, objects or functions. During each rendering cycle, deck.gl shallow-compares them with the previous values.

After reading above in documents, i understand that the value of accessor of layer will be updated whenever the variables in `updateTriggers` change.

By the way  i have updated the new sandbox code based on my current understanding, i use the button at the bottom of page to change the `color` value provided in the `updateTriggers` `getLineColor` but it does not work.

**Reply by Pessimistress** [reply]
As I said, this is not about your understanding of deck.gl. Your component will not rerender by reassigning a variable. You need to read some React tutorials.

**Reply by Pessimistress** [reply]
```js
function App() {
  const [color, setColor] = useState([0, 0, 0]);
  const layer = new GeoJsonLayer({
    ...
    getLineColor: f => ...
    updateTriggers: {
      getLineColor: color
    }
  });
  const changeColor = () => {
    setColor([255, 0, 0]);
  };

  return <DeckGL initialViewState={...} controller={true} layers={[layer]}>
    <button onClick={changeColor} >change color</button>
  </DeckGL>
}
```

**Reply by hoang-ta** [reply]
I understand that you have to use `color ` as state so that when `setColor` app will re-render then layer will use new `color` value.

Actually,  what i need to understand  is **what is the use of  `updateTriggers`**,  because currently, with or without `updateTriggers`, it does not affect the result when clicking a button.

I have updated the [sandbox](https://codesandbox.io/s/jovial-pare-dlmqo?file=/src/App.tsx) code that look quite similar to deck.gl document (https://deck.gl/docs/api-reference/core/layer#updatetriggers)

**Reply by Pessimistress** [reply]
You are using a new `data` object on every render. If `data` changes shallowly then all accessors are called again. `updateTriggers` is only useful if you don't change the `data` object for performance reasons. See https://deck.gl/docs/developer-guide/performance#use-updatetriggers

**Reply by hoang-ta** [reply]
Thank you for pointing this out for me, I put data outside App then `updateTriggers` works as expected.
